File: src/gmp_test/src/optimization/constr_opt/mpc/gmpMpcOpt.cpp
#include <iostream>
#include <algorithm>

#include <gmp_lib/GMP/GMP_MPC.h>
#include <gmp_lib/io/gmp_io.h>

#include <ros/package.h>

using namespace as64_;

void PRINT_INFO_MSG(const std::string &msg, std::ostream &out=std::cout);
void PRINT_CONFIRM_MSG(const std::string &msg, std::ostream &out=std::cout);
void PRINT_WARNING_MSG(const std::string &msg, std::ostream &out=std::cerr);
void PRINT_ERROR_MSG(const std::string &msg, std::ostream &out=std::cerr);

int main(int argc, char **argv)
{
  // data path
  std::string path = ros::package::getPath("gmp_test") + "/src/optimization/constr_opt/mpc/data/";

  // read input data
  arma::vec y0;
  arma::vec yg;
  double tau;
  arma::mat pos_lim;
  arma::mat vel_lim;
  arma::mat accel_lim;
  gmp_::GMP::Ptr gmp(new gmp_::GMP());
  {
    gmp_::FileIO fid(path + "gmp_mpc_opt_in.bin", gmp_::FileIO::in);
    fid.read("y0", y0);
    fid.read("yg", yg);
    fid.read("tau", tau);
    fid.read("pos_lim", pos_lim);
    fid.read("vel_lim", vel_lim);
    fid.read("accel_lim", accel_lim);
    gmp_::read(gmp.get(), fid, "gmp_");
  }

  unsigned n_dof = y0.size();
  arma::vec ones_ndof = arma::vec().ones(n_dof);
  arma::vec O_ndof = arma::vec().zeros(n_dof);

  arma::rowvec Time;
  arma::mat P_data;
  arma::mat dP_data;
  arma::mat ddP_data;
  arma::mat slack_data;

  double t = 0;
  double dt = 0.002;
  double s = t/tau;
  double s_dot = 1/tau;
  double s_ddot = 0;
  arma::vec y = y0;
  arma::vec y_dot = O_ndof;
  arma::vec y_ddot = O_ndof;

  // --------  GMP - MPC  --------
  int N_horizon = 12;
  double pred_time_step = 0.02;
  int N_kernels = 30;
  double kernels_std_scaling = 1.5;

  bool pos_slack = 1;
  bool vel_slack = 1;
  bool accel_slack = 1;
  std::vector<bool> slack_flags = {pos_slack, vel_slack, accel_slack};
  std::vector<double> slack_gains = {1e6, 100, 20};

  gmp_::GMP_MPC gmp_mpc(gmp.get(), N_horizon, pred_time_step, N_kernels, kernels_std_scaling, slack_flags, slack_gains);

  gmp_mpc.settings.time_limit = 1.5e-3;
//  gmp_mpc.settings.max_iter = 4000;
  gmp_mpc.settings.abs_tol = 1e-3;
  gmp_mpc.settings.rel_tol = 1e-4;

  gmp_mpc.setPosLimits(pos_lim.col(0), pos_lim.col(1));
  gmp_mpc.setVelLimits(vel_lim.col(0), vel_lim.col(1));
  gmp_mpc.setAccelLimits(accel_lim.col(0), accel_lim.col(1));

  gmp_mpc.setPosSlackLimit(1e-3);
  gmp_mpc.setVelSlackLimit(0.02);
  gmp_mpc.setAccelSlackLimit(0.2);

  gmp_mpc.setInitialState(y, y_dot, y_ddot, s, s_dot, s_ddot);
  gmp_mpc.setFinalState(yg, O_ndof, O_ndof, 1, 0, 0);

  gmp->setScaleMethod( gmp_::TrajScale::Ptr( new gmp_::TrajScale_Prop(n_dof) ) );
  gmp->setY0(y0);
  gmp->setGoal(yg);

  arma::rowvec elaps_t_data;

  arma::wall_clock timer, timer2;
  timer.tic();
  // --------  Simulation loop  --------
  while (true)
  {
    // --------  Stopping criteria  --------
    if (s > 1.0) break;

    //text_prog.update(100*t/tau);

    // std::cout << 100*t/tau << "\n";

    if (s >= 1)
    {
      s = 1;
      s_dot = 0;
      s_ddot = 0;
    }

    if (s > 0.9)
    {
      gmp_mpc.settings.abs_tol = 1e-4;
      gmp_mpc.settings.rel_tol = 1e-5;
    }


    timer2.tic();
    gmp_::GMP_MPC::Solution sol = gmp_mpc.solve(s, s_dot, s_ddot);
    arma::vec y = sol.y;
    arma::vec y_dot = sol.y_dot;
    arma::vec y_ddot = sol.y_ddot;
    arma::vec slack_var = sol.slack_var;
    elaps_t_data = arma::join_horiz( elaps_t_data, arma::vec({timer2.toc()*1000}) );

    // gmp_mpc.setInitialState(y, y_dot, y_ddot, s, s_dot, s_ddot);

    if (sol.exit_flag > 0) PRINT_WARNING_MSG(sol.exit_msg + "\n");
    else if (sol.exit_flag < 0)
    {
      PRINT_ERROR_MSG(sol.exit_msg + "\n");
      break;
    }

    // --------  Log data  --------
    Time = arma::join_horiz(Time, arma::mat({t}));
    P_data = arma::join_horiz(P_data, y);
    dP_data = arma::join_horiz(dP_data, y_dot);
    ddP_data = arma::join_horiz(ddP_data, y_ddot);
    slack_data = arma::join_horiz(slack_data, slack_var);

    // --------  Numerical integration  --------
    t = t + dt;
    s = s + s_dot*dt;
    s_dot = s_dot + s_ddot*dt;
  }

  //text_prog.update(100);

  double elaps_t_ms = timer.toc()*1000;
  PRINT_INFO_MSG("===> GMP-MPC optimization finished! Elaps time: " + std::to_string(elaps_t_ms) + " ms\n");

  double mean_elaps_t = arma::mean(elaps_t_data);
  double std_elaps_t = arma::stddev(elaps_t_data);
  double max_elaps_t = arma::max(elaps_t_data);
  double min_elaps_t = arma::min(elaps_t_data);

  std::cout << "======= Elapsed time (ms) ======\n";
  std::cout << "std_range: [" << std::max(0.,mean_elaps_t-std_elaps_t) << " -  " << mean_elaps_t + std_elaps_t <<"]\n";
  std::cout << "mean     : " << mean_elaps_t << " +/- " << std_elaps_t <<"\n";
  std::cout << "min      : " << min_elaps_t << "\n";
  std::cout << "max      : " << max_elaps_t << "\n";
  std::cout << "==============================\n";

  // ----------- write results ----------------
  {
      gmp_::FileIO fid(path + "gmp_mpc_opt_out.bin", gmp_::FileIO::out | gmp_::FileIO::trunc);
      fid.write("Time", Time);
      fid.write("P_data", P_data);
      fid.write("dP_data", dP_data);
      fid.write("ddP_data", ddP_data);
      fid.write("slack_data", slack_data);
      fid.write("pos_slack", pos_slack);
      fid.write("vel_slack", vel_slack);
      fid.write("accel_slack", accel_slack);
  }


  return 0;
}

void PRINT_INFO_MSG(const std::string &msg, std::ostream &out)
{
  out << "\033[1m" << "\033[34m" << "[INFO]: " << msg << "\033[0m";
}

void PRINT_CONFIRM_MSG(const std::string &msg, std::ostream &out)
{
  out << "\033[1m" << "\033[32m" << "[INFO]: " << msg << "\033[0m";
}

void PRINT_WARNING_MSG(const std::string &msg, std::ostream &out)
{
  out << "\033[1m" << "\033[33m" << "[WARNING]: " << msg << "\033[0m";
}

void PRINT_ERROR_MSG(const std::string &msg, std::ostream &out)
{
  out << "\033[1m" << "\033[31m" << "[ERROR]: " << msg << "\033[0m";
}

File: src/gmp_test/src/optimization/constr_opt/mpc/gmpMpcOpt.m
function [Time, P_data, dP_data, ddP_data] = gmpMpcOpt(gmp0, tau, y0, yg, pos_lim, vel_lim, accel_lim, opt_pos, opt_vel, qp_solver_type)
      
    gmp = gmp0.deepCopy();
   
    n_dof = length(y0);
    
    Time = [];
    P_data = [];
    dP_data = [];
    ddP_data = [];
    slack_data = [];
    
    O_ndof = zeros(n_dof,1);

    t = 0;
    dt = 0.005;
    s = t/tau;
    s_dot = 1/tau;
    s_ddot = 0;
    y = y0;
    y_dot = O_ndof;
    y_ddot = O_ndof;
    
    slack_gains = [1e5 100 1];
    
    %% --------  GMP - MPC  --------
    gmp_mpc = GMP_MPC(gmp, 10, 0.1, 30, 1.5, opt_pos, opt_vel, slack_gains); % try 12 , 0.025
    
    gmp_mpc.setPosLimits(pos_lim(:,1), pos_lim(:,2));
    gmp_mpc.setVelLimits(vel_lim(:,1), vel_lim(:,2));
    gmp_mpc.setAccelLimits(accel_lim(:,1), accel_lim(:,2));
    
    gmp_mpc.setPosSlackLimit(5e-3);
    gmp_mpc.setVelSlackLimit(0.05);
    gmp_mpc.setAccelSlackLimit(0.2);
    
%     gmp_mpc.setPosSlackLimit(0);
%     gmp_mpc.setVelSlackLimit(0);
%     gmp_mpc.setAccelSlackLimit(0);
    
    gmp_mpc.setInitialState(y, y_dot, y_ddot, s, s_dot, s_ddot);
    gmp_mpc.setFinalState(yg, O_ndof, O_ndof, 1, s_dot, 0, 1e-3*ones(n_dof,1));
    
    gmp.setScaleMethod(TrajScale_Prop(n_dof));
    gmp.setY0(y0);
    gmp.setGoal(yg);
    
    text_prog = ProgressText(40);
    text_prog.init();
    
    t_start = tic;

    %% --------  Simulation loop  --------
    while (true)
        
        %% --------  Stopping criteria  --------
        %if (s > 1.0), break; end
        if (s >= 1 && norm(y_dot)<1e-2 && norm(y-yg)<1e-2 ), break; end

        if (s<=1), text_prog.update(100*t/tau); end

%         if (s >= 1)
%             s = 1;
%             s_dot = 0;
%             s_ddot = 0;
%         end     

        [exit_flag, y, y_dot, y_ddot, slack_var] = gmp_mpc.solve(s, s_dot, s_ddot);
        
        % gmp_mpc.setInitialState(y, y_dot, y_ddot, s, s_dot, s_ddot);

        if (exit_flag)
            warning(gmp_mpc.getExitMsg());
            text_prog.printInNewLine();
            if (exit_flag < 0), break; end
        end  

        %% --------  Log data  --------
        Time = [Time t];
        P_data = [P_data y];
        dP_data = [dP_data y_dot];
        ddP_data = [ddP_data y_ddot];

        slack_data = [slack_data slack_var];

        %% --------  Numerical integration  --------
        t = t + dt;
        s = s + s_dot*dt;
        s_dot = s_dot + s_ddot*dt;
    
    end

    if (~isempty(slack_data))
        
        pos_slack = slack_gains(1) > 0;
        vel_slack = slack_gains(2) > 0;
        accel_slack = slack_gains(3) > 0;
        
        n_slack = size(slack_data,1);
        y_lb = {};
        if (pos_slack), y_lb = [y_lb, {'pos'}]; end
        if (vel_slack), y_lb = [y_lb, {'vel'}]; end
        if (accel_slack), y_lb = [y_lb, {'accel'}]; end
        figure;
        for i=1:n_slack
            subplot(n_slack,1,i);
            plot(Time, slack_data(i,:), 'LineWidth',2, 'Color','red');
            ylabel(y_lb{i}, 'fontsize',15);
            if (i==1), title('slack variables', 'fontsize',17); end
        end
    end
    
    text_prog.update(100);
    fprintf('\n');
    fprintf('===> GMP-MPC optimization finished! Elaps time: %f ms\n',toc(t_start)*1000);

end
